proposals/NNNN-task-cancellation-shields.md

```diff
@@ -0,0 +1,340 @@
+  }
+}
+```
+
+It is meaningless to try to shield the `addTask` operation of a task group as it does not enclose the lifetime or any part of the child tasks execution. Instead you should shield the child task within the `addTask` function if shielding a specific task is your goal:
+
+```swift
+await withDiscardingTaskGroup { group in 
+  // ❌ has no effect on child task observing cancellation:
+  withTaskCancellationShield { 
+    group.addTask { ... } 
+  } 
+  
+  
+  // 🟢 does properly shield specific child task observing cancellation:
+  group.addTask { 
+    withTaskCancellationShield { ... }
+  } 
+}
+```
+
+All examples shown using `isCancelled` behave exactly the same for `Task.checkCancellation`, i.e. whenever `isCancelled` would be true, the `checkCancelled` API would throw a `CancellationError`.
+
+### Cancellation Shields and Cancellation Handlers
+
+Swift concurrency offers task cancellation handlers which are invoked immediately when a task is cancelled. This allows you to dynamically react to cancellation happening without explicitly checking the `isCancelled` property of a task. 
+
+Task cancellation shields also prevent cancellation handlers from firing if the handler was stored while a shield was active. Again, this does not extend to child tasks, but only to the current task that is being shielded. 
+
+For example, the task cancellation shield installed around the `slowOperation` in the snippet below, would effectively prevent the cancellation handler inside the `slowOperation` function from ever triggering:
+
+```swift
+func slowOperation() -> ComputationResult {
+  await withTaskCancellationHandler { 
+    return < ... slow operation ... >
+  } onCancel: {
+    print("Let's cancel the slow operation!")
+  }
+}
+
+func cleanup() {
+  withTaskCancellationShield {
+    slowOperation()
+  }
+}
+```
+
+### Cancellation Shields and Task handles
+
+Unstructured tasks, as well as the use of `withUnsafeCurrentTask`, offer a way to obtain a task handle which may be interacted with outside of the task.
+
+For example, you may obtain a task handle a an unstructured task, which then immediately enters a task cancellation shield scope:
+
+```swift
+let task = Task { 
+  Task.isCancelled // true
+  withTaskCancellationShield { 
+    Task.isCancelled // false
+  }
+  Task.isCancelled // true
+}
+
+task.cancel()
+print(task.isCancelled) // _always_ true
+```
+
+The **instance method** `task.isCancelled` queried from the outside of the task will return the _actual_ cancelled state, regardless if the task is right no executing a section of code under a cancellation shield or not. This is because from the outside it would be racy to query the cancellation state and rely on wether or not the task is currently executing a section of code under a shield. This could lead to confusing behavior where querying the same `task.isCancelled` could be flip flopping between cancelled and not cancelled.
```

Comment:
```suggestion
The **instance method** `task.isCancelled` queried from the outside of the task will return the _actual_ cancelled state, regardless of whether the task is currently executing a section of code under a cancellation shield or not. This is because from the outside it would be racy to query the cancellation state and rely on whether or not the task is currently executing a section of code under a shield. This could lead to confusing behavior where querying the same `task.isCancelled` could be flip flopping between cancelled and not cancelled.
```